### DISCUSSION TITLE
"Songlink" plug-in

### DISCUSSION THREAD

**Discussion by ChipSendype** [audacious-media-player/audacious#1534] [Ideas]
Hello ya'll, 
I actually joined Github just to post this. I am a construction guy (Builder, heavy equipment, etc.). 
In relation to most Github folks, I barely know what I'm doing on a computer. 
I just started using Linux (Mint) in 2020 and have never written any code or scripts...etc.
I'm really good at cut and paste though. :)  
Just to give you a little context on why I don't just create something myself. 
I even tried using AI to help walk me through the process. AI isn't there yet for idiots like me.

With that being said, here's my idea: I did search for other similar ideas. Nothing to be found.
I'm old (school), 54. Some songs absolutely MUST be played back to back, to include 
intro's, intermissions, transitional songs...etc. There are literally thousands of songs I can think of that are like this and it's a wicked bummer to have a terrible segway from what one is used to hearing, and what is played while in shuffle mode. 
It would be really cool if there was a way to link two tracks together that over-rides shuffle 
temporarily. The feature would simply turn shuffle off at the beginning of the chosen track using metadata, then, the next sequential track would turn shuffle back on using metadata. 
In this way, while in shuffle mode, those two songs will play together. 

Given my research, one may possibly have to create an entirely new class of ID3 tag (SHUF)? in order for the feature to apply to any device that plays mp3's ie. car stereo, mp3 player, etc. 
Not sure if it's an easier task to make this feature only for Audacious. 
I think one still has to be able to assign control data to the ID3v2 metadata frame, so I think this would have to be done through an mp3 tag editor like 'easyTAG'.  
I could not identify the binary code for controlling the shuffle, where it is inserted and if that trigger is controllable through ID3. 

Regardless, this is waaay over my head. I need a nap every time I think about it.  
At the very least, I think this would be a great new feature that has been long overdue in digital music players. 
Audacious is my go-to player and for me, it's the best. Thanks to everyone that contributes, I appreciate you all. Happy Holidays.

**Comment by MikeSS8** [comment]
>Some songs absolutely MUST be played back to back, to include
intro's, intermissions, transitional songs...etc. There are literally thousands of songs I can think of that are like this and it's a wicked bummer to have a terrible segway from what one is used to hearing, and what is played while in shuffle mode.

Seems simple enough: concatenate the songs that should always play in a particular order.

Using FFmpeg, for example:

```sh
ffmpeg -f concat -safe 0 -i <(printf "file '%s'\n" prelude.mp3 intermezzo.mp3 finale.mp3) thewholesequence.flac
```

**Reply by ChipSendype** [reply]
Hi, thanks for taking the time to respond. 
For me, what you shared may as well be Chinese.  I suppose not defining what I want more clearly, is my bad. 
Allow me to be more specific in what I would like to see for this feature. 
Remember, I am a newborn baby, an idiot. I move dirt for a living. 
So, the implementation would be either linking inside a GUI like easyTAG, or Audacious. 
Only by GUI...you know, for dum dums like me. :) 
I don't see how using FFmpeg would effect playback on a car stereo for example, unless 
FFmpeg does indeed alter the ID3 tag frames. 
Considering this, I imagine the best course of action is using a tag frame that toggles 
shuffle on/off regardless of where that mp3 goes. 
Then it's simply a matter of how that would be executed/represented inside the easyTAG or Audacious GUI.

**Reply by mschwendt** [reply]
> I don't see how using FFmpeg would effect playback on a car stereo for example,

The suggested command uses ffmpeg to concatenate the individual tracks you want to play consecutively. You would take the resulting audio file and play it in your car without any problems.

Apart from that, it doesn't need to be ffmpeg and command-line tool usage. You could concatenate audio files also with a GUI audio editing program of your choice. Keep the original files somewhere safe, listen to the edited/concatenated files in exactly the way you want it.

> unless FFmpeg does indeed alter the ID3 tag frames.

That wouldn't work, because you cannot simply introduce new ID3 tags and hope that hardware & software support those tags "magically". Especially hardware audio player implementations would not know about those new tags. But the whole idea of inserting cross-file links within audio files would be highly problematic already.

**Reply by ChipSendype** [reply]
Thanks again. I guess it will remain a wishful thought. 
I had to actually look up the word, "concatenate". By next week, I will not know what it means again. 
I still appreciate your time and effort.  Happy New Year.

**Comment by wildstar84** [comment]
I think he meant to suggest simply creating a new song file containing both songs as one, but you are correct that the ID3 tags would not change to those for the 2nd song in the middle of playing the new, combined song file, but rather would likely have the tags of the first of the 2 songs, which is not what you (or anyone else would want).  Also, I wouldn't just assume this will remain only a wish, just b/c the replyer hasn't committed Audacious to implement it.

**Reply by ChipSendype** [reply]
Thanks for the response, Wildstar84.
I'm so new to all this, Github, Linux, owning my own mp3's...etc.
I think it's a good idea and I have done a little research. Well, to me it seems like a lot of research, but
not knowing what the heck I'm reading more than half the time doesn't really count. :)
But I'll take a solid B+ for persistence. 
I realize the way music is being consumed these days, that "Songlink" may not appeal to many younger generations, but for the minority that still listens to whole albums, or remembers songs that absolutely go together (Led Zeppelin has a lot of these), I still think it's a valuable idea/feature for true music lovers.
I'm only an idea man that moves dirt and rocks around, I don't speak computer too goodly. 
I would love to see the idea more professionally sussed out. 
I joined id3.org recently to explore more of that end of it. To see if player control is even possible
using frame metadata and hence, the creation of a new frame category (SHUF).
Happy New Year!

**Reply by mschwendt** [reply]
Cassette players didn't offer "shuffle" mode as a feature. Vinyl record players not either. ;-)

When I created digital copies of my old cassette tapes and vinyl records I decided myself which tracks to append to eachother.


Anyway, no other audio player would understand a modified/hacked ID3. And overriding player's shuffle setting would not be accepted as a standard in my opinion.

**Reply by ChipSendype** [reply]
Thanks for the input mschwendt.  
Perhaps you're right, but stranger things have happened and I don't see any other way to accomplish the 
goal without toggling shuffle off and on or, as you mentioned, editing mp3's together. Considering the vast majority of people aren't going to edit hundreds of songs together in Audacity, and hardware and software devices use the same binary triggers for shuffle as well as the same Fisher-Yates algorithm for shuffle, I would bet it would be more ubiquitous/applicable than not.
It's not like a permanent override, just a momentary toggle off and on, hardly noticeable by the user. 
I also am not talking about "modifying or hacking" an ID3. I'm talking about creating an entirely new ID3 frame (SHUF) who's sole designed purpose is toggling shuffle off and on. No different than any other ID3 frame category in which mp3's function perfectly. 
The only wide variable is how it would be executed in an mp3 editing GUI or player. Each team of that software would have to figure out their particular GUI implementation.

**Reply by mschwendt** [reply]
> Considering the vast majority of people aren't going to edit hundreds of songs together in Audacity,

They wouldn't edit their existing MP3 files either. Literally billions of audio files copied between devices.

> and hardware and software devices use the same binary triggers for shuffle as well as the same Fisher-Yates algorithm for shuffle,

Likely not. I've had audio players that choose the next track on the list randomly on-the-fly, without supporting going forward/backward through a fully randomized playlist history.

>I'm talking about creating an entirely new ID3 frame (SHUF) who's sole designed purpose is toggling shuffle off and on.

Extremely questionable for a multitude of reasons. Already if you think about it at the low level of shuffling a list, by the time you create the randomized playlist positions, you break your original sequence of files that override the player's shuffle setting.

Not that I want to encourage you to do that, but you could always join the mailing-list at https://id3.org
